Enhance default detection

pdanpdan opened the pull request:
- process default on all structure types
- process complex array types (array of arrays/enums/objects)
- add tests for 100% coverage

pdanpdan commented:
As this changes parameters for the internal functions I made the PR based on the https://github.com/thedumbterminal/jsonschema-avro/pull/25 PR.
If there are changes required for https://github.com/thedumbterminal/jsonschema-avro/pull/26 I'll rebase this after the merge to avoid problems.

thedumbterminal commented:
If this change is still required can we isolate this PR to just deal with that feature and any other changes to their specific PRs. Thanks for the work on this.

pdanpdan committed:
Enhance default detection

- process default on all structure types
- process complex array types (array of arrays/enums/objects)
- process root level arrays
- add tests for 100% coverage

pdanpdan commented:
Hello
- I have updated/rebased the PR
- I have added more tests
- the only changes to existing tests are:
  - in omit_properties/expected.json I also removed the top level `doc` property to have one more test
  - in all expected outputs I removed the empty `doc` property (which should not be there)

thedumbterminal commented on src/index.js:
```diff
@@ -82,142 +76,250 @@ jsonSchemaAvro._isArray = (schema) => schema.type === 'array'
+  return mappedType
+}
+
+jsonSchemaAvro._setTypeAndDefault = (
```
I like this approach!

thedumbterminal reviewed (commented):
Cheers, will take a look through this when I have a moment, thanks

pdanpdan commented:
No changes, just a rebase to check that all tests pass after change from float to double

thedumbterminal reviewed (approved):
Nice! Thanks for taking the time to put together a comprehensive change
